### PR TITLE
Refactor some of  `dereference.rs` to util functions

### PR DESCRIPTION
**Pull request by Jarcho**
I've seen a few lints that need to be able to tell if changing the type of an expression would be a vaild suggestion. This extracts part of how that's done from `explicit_auto_deref`.

changelog: None

### CODE REVIEW

File: clippy_utils/src/lib.rs [hunk 1]
```diff
@@ -2502,6 +2505,263 @@ pub fn walk_to_expr_usage<'tcx, T>(
...
+    Field(&'tcx ExprField<'tcx>),
+    /// An argument to a function.
+    FnArg(&'tcx Expr<'tcx>, usize),
+    /// An argument to a method.
+    MethodArg(HirId, Option<&'tcx GenericArgs<'tcx>>, usize),
+    /// The callee of a function call.
+    Callee,
+    /// Access of a field.
+    FieldAccess(Ident),
+}
```
Comment (smoelius) [inline clippy_utils/src/lib.rs:2548]:
```suggestion
}

```

File: clippy_lints/src/dereference.rs [hunk 2]
```diff
@@ -649,420 +788,164 @@ fn deref_method_same_type<'tcx>(result_ty: Ty<'tcx>, arg_ty: Ty<'tcx>) -> bool {
...
-        matches!(
-            self,
-            Self::MethodReceiver | Self::FieldAccess { of_union: false, .. } | Self::Callee
-        )
+#[derive(Clone, Copy)]
+enum TyCoercionStability {
+    Deref,
+    Reborrow,
+    None,
+}
```
Comment (smoelius) [inline clippy_lints/src/dereference.rs:818]:
```suggestion
}

```
Is juxtaposing the the type and its impl intentional?

Comment (Centri3) [reply clippy_lints/src/dereference.rs:818]: `blank_lines_lower_bound` is definitely relevant here. Is it worth the conflicts though? :eyes:

Comment (Jarcho) [reply clippy_lints/src/dereference.rs:818]: Formatting the type and it's impl without a blank line in between is used in other parts of clippy. It's not a consistent thing though.

**Comment by smoelius** [comment]
> I've seen a few lints that need to be able to tell if changing the type of an expression would be a vaild suggestion. This extracts part of how that's done from `explicit_auto_deref`.

Can I ask how/where you see the extracted code being applied?

File: clippy_lints/src/dereference.rs (hunk 2 above)
Comment (smoelius) [reply clippy_lints/src/dereference.rs:818]: Are the [multiple impls](https://github.com/rust-lang/rust-clippy/blob/e9f65e3315c068505bda195784d89a58b14e8dc3/clippy_utils/src/lib.rs#L2549-L2561) similarly intentional?

Comment (Jarcho) [reply clippy_lints/src/dereference.rs:818]: That part isn't.

**Comment by smoelius** [comment]
Thanks!

File: clippy_utils/src/lib.rs [hunk 3]
```diff
@@ -2502,6 +2505,261 @@ pub fn walk_to_expr_usage<'tcx, T>(
...
+    pub fn is_return(&self) -> bool {
+        matches!(self, Self::Return(_))
+    }
+
+    /// Checks if the value is used as a method call receiver.
+    pub fn is_recv(&self) -> bool {
+        matches!(self, Self::MethodArg(_, _, 0))
+    }
+
+    /// Get's the needed type as it's defined without any type inference.
```
Comment (Centri3) [inline clippy_utils/src/lib.rs:2560]:
```suggestion
    /// Gets the needed type as it's defined without any type inference.
```

File: clippy_utils/src/lib.rs [hunk 4]
```diff
@@ -2502,6 +2505,261 @@ pub fn walk_to_expr_usage<'tcx, T>(
     None
 }
 
+/// A type definition as it would be viewd from within a function.
```
Comment (Centri3) [inline clippy_utils/src/lib.rs:2508]:
```suggestion
/// A type definition as it would be viewed from within a function.
```

**Pull request merged**